### PR TITLE
Fastfilter enhancements

### PR DESCRIPTION
**Pull request by gandrewstone**
Implement filters with K (# hashes) between 2 and 16.  Add a salt to prevent hash collision attacks.

Currently many filters are configured for m=4 million bits.  With k=16 and n=100000 transactions, this results in a false positive rate of  1.94 × 10^-8, based on the FP rate equation (1-e^(-kn/m))^k.

### CODE REVIEW

File: src/fastfilter.h [hunk 1]
```diff
@@ -43,47 +45,74 @@ class CFastFilter
 
     CFastFilter()
     {
+        static_assert((HASHES > 1) && (HASHES <= 16), "HASHES must be between 2 and 16 inclusive");
```
Comment (awemany) [inline src/fastfilter.h]: Why two as minimum?

File: src/fastfilter.h [hunk 2]
```diff
@@ -43,47 +45,74 @@ class CFastFilter
...
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
+            uint32_t idx2 = val & (FILTER_SIZE - 1);
+
+            vData[idx >> 3] |= (1 << (idx & 7));
+            vData[idx2 >> 3] |= (1 << (idx2 & 7));
+        }
```
Comment (awemany) [inline src/fastfilter.h:105]: The hash calculation is code duplication and should IMO be replaced with an `inline` function. Especially since it grew like this.

To be honest, I am not yet convinced of the whole fast filter approach being worthwhile yet. I really like to see benchmarks!

File: src/fastfilter.h [hunk 3]
```diff
@@ -43,47 +45,74 @@ class CFastFilter
...
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
+            uint32_t idx2 = val & (FILTER_SIZE - 1);
+
+            unset |= (0 == (vData[idx >> 3] & (1 << (idx & 7))));
+            unset |= (0 == (vData[idx2 >> 3] & (1 << (idx2 & 7))));
+        }
+        return !unset;
```
Comment (awemany) [inline src/fastfilter.h:122]: More code dup.

File: src/fastfilter.h [hunk 4]
```diff
@@ -95,7 +124,8 @@ class CRollingFastFilter : public CFastFilter<FILTER_SIZE>
         erase = insecure_rand.rand32() % CFastFilter<FILTER_SIZE>::FILTER_BYTES;
         eraseAmt = _eraseAmt;
     }
-    void insert(const uint256 &hash)
+
+    void roll(void)
```
Comment (awemany) [inline src/fastfilter.h:143]: Has this `roll` approach been tested to be sufficiently unbiased?

File: src/test/fastfilter_tests.cpp [hunk 5]
```diff
@@ -37,45 +37,51 @@ BOOST_AUTO_TEST_CASE(fastfilter_tests)
             BOOST_CHECK(filt.contains(tmp));
             BOOST_CHECK(!filt.checkAndSet(tmp));
         }
-        BOOST_CHECK(collisions < 400); // sanity check, actual result may vary
+        BOOST_CHECK(collisions < 10); // sanity check, actual result may vary
```
Comment (awemany) [inline src/test/fastfilter_tests.cpp:40]: How often (about) will this (and the similar narrow tests) fail sporadically?

File: src/fastfilter.h [hunk 6]
```diff
@@ -27,13 +27,15 @@ class uint256;
  * but "inserts" may be lost.  However, if you are using this class as an in-ram filter before doing a more expensive
  * operation, a lost insert may be acceptable.
  */
-template <unsigned int FILTER_SIZE>
+template <unsigned int FILTER_SIZE, unsigned int HASHES = 16>
```
Comment (awemany) [inline src/fastfilter.h]: If I understand you correctly, this is hash *functions* instead of *hashes* and thus the name should be changed.

File: src/fastfilter.h [hunk 7]
```diff
@@ -43,47 +45,74 @@ class CFastFilter
...
-        vData[idx >> 3] |= (1 << (idx & 7));
-        if (ret)
-            conflicts++;
-        return !ret;
+        const uint32_t *saltPos = (const uint32_t *)salt.begin();
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        bool unset = 0; // If any position is not set, then this will be true
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
```
Comment (awemany) [inline src/fastfilter.h]: Does XORing really help here with avoiding collisions? What is the scenario that it avoids?

File: src/fastfilter.h [hunk 8]
```diff
@@ -43,47 +45,74 @@ class CFastFilter
...
-        const uint32_t *pos = (const uint32_t *)&(mem[grabFrom]);
-        uint32_t idx = ((*pos) ^ mem[0]) & (FILTER_SIZE - 1);
-        vData[idx >> 3] |= (1 << (idx & 7));
+        const uint32_t *saltPos = (const uint32_t *)salt.begin();
+        const uint32_t *pos = (const uint32_t *)hash.begin();
+        for (unsigned int i = 0; i < HASHES / 2; i++, pos++, saltPos++)
+        {
+            uint32_t val = *pos ^ *saltPos;
+            uint32_t idx = val & (FILTER_SIZE - 1);
+            val = __builtin_bswap32(val);
```
Comment (awemany) [inline src/fastfilter.h:100]: What is the reason for the swap?

EDIT: As per the discussion this is for getting more bits out of this 32-bit value, but which will reuse bits when using more than 16! I think an approach that would be more amenable to proper analysis would be to just properly slice up the input hash into log2(FILTER_SIZE-1) bits each.

**Review by awemany — CHANGES_REQUESTED** [review]

**Comment by awemany** [comment]
Another general comment here: It looks like there's no difference between the fast filter behavior for an even `n` number of hash functions and the odd `n+1` number of hash functions. In that sense, the fast filter is ignoring user input a bit. Minor nit maybe.

Then, I think it would make sense to remove the XOR "salt". Or maybe replace it with a permutation table - doing byte-wise permutation, this should make it hard for an adversary to reduce the effective number of hash functions by 'mining' TXIDs when the number of hash functions is very small. (Idea being that it would make it `log2(factorial(32)) ~= 118 bits` harder to select any particular pattern.)

If degrading performance for the full `n=16` case is a worry, a separate code path should be easy enough for the compiler to select and optimize.

This said, I think the approach needs a little bit more analysis but the more I think about it the more I suspect it will actually be fine also under adversarial conditions.

EDIT: And another comment on the `roll()` method. In the 'rolling' case this rather becomes an approximate set structure, correct? I wonder whether it would make sense to randomly erase a location upon any insert as this seems to me to reflect the necessary steady state case where an insert will not change the overall FPR and FNR of this approximate set. And I suspect *randomly* deleting will be strictly better (also in adverserial conditions) rather than deleting with an increasing position counter.

File: src/fastfilter.h (hunk 1 above)
Comment (gandrewstone) [reply src/fastfilter.h]: The loop is unrolled to do 2 at once

File: src/fastfilter.h (hunk 2 above)
Comment (gandrewstone) [reply src/fastfilter.h:105]: I want to rely as little as possible on compiler optimization, since its easy for such to silently disappear based on small tangentially related changes.

File: src/fastfilter.h (hunk 3 above)
Comment (gandrewstone) [reply src/fastfilter.h:122]: ditto

File: src/fastfilter.h (hunk 4 above)
Comment (gandrewstone) [reply src/fastfilter.h:143]: since the input hashes are unbiased, they are placed randomly within the bloom filter.  therefore a sequential roll is unbiased.  I'm not sure exactly what kind of test you have envisioned but you should look at what I have in the fastfilter_tests.cpp file.

File: src/test/fastfilter_tests.cpp (hunk 5 above)
Comment (gandrewstone) [reply src/test/fastfilter_tests.cpp:40]: The following python code makes the assumption that each test is independent.  This is not quite true since a prior collision means that fewer bits will be actually set in the data, resulting in a slightly smaller probability of subsequent collision.

```import math

def bloomCollision(n,m, k):
    return (1.0 - math.e**(-1*float(k)*float(n)/float(m)))**k

def Test():
    probs= []
    for i in range(50000):
        probs.append(bloomCollision(i,1000000, 16))

    s = sum(probs)
    m = 1.0
    for i in probs:
        m *= i
    oneCollision = s-m
    tenCollision = oneCollision**10
    print(tenCollision)
```

So looks like 6.430521732653953e-06

File: src/fastfilter.h (hunk 6 above)
Comment (gandrewstone) [reply src/fastfilter.h]: ok

File: src/fastfilter.h (hunk 8 above)
Comment (gandrewstone) [reply src/fastfilter.h:100]: this is a single instruction on an intel machine.  It is OK to reuse bits.  Consider that the traditional bloom filter reuses all the input bits in every hash function.

**Comment by gandrewstone** [comment]
I have added a performance test.  First, the bitcoin "bench" seems pretty inaccurate for a variety of reasons beyond the scope of this comment so bench results should be taken loosely.  In this performance test, I first approximately matched the memory and # hashes for the bloom and fastfilter.  Next, I ran fastfilter with 2 hash function to get a maximum performance value.

I measured 2 common operations -- check for membership and insert if it does not match, and just check.

For "check and insert" operations, fastfilter performs approximately 20 times faster than the bloom filter with comparable sizes, and 70 times faster with 2 hash functions.

For "check" operations, fastfilter performs 10x faster than the comparable bloom filter and 35x faster with 2 hash functions.  Basically,  the difference between "check and insert" and "check" can be explained because the fastfilter has an optimized "check and insert" function that avoids accessing memory multiple times.

```
BloomCheckSet,1535,0.000635982,0.00069198,0.000664231
BloomContains,3071,0.000333363,0.000905991,0.000337465
FastFilterCheckSet,26623,3.67463e-05,4.10175e-05,4.01079e-05
FastFilterCheckSet2,106495,8.46386e-06,1.00955e-05,9.52414e-06
FastFilterContains,26623,3.92795e-05,0.000144958,4.01817e-05
FastFilterContains2,106495,9.28342e-06,2.89679e-05,9.94415e-06
```

File: src/fastfilter.h (hunk 2 above)
Comment (awemany) [reply src/fastfilter.h:105]: How about I rework this to use `inline` and you try to find such a small tangentially related change that will screw it up? :)

Compilers are quite good nowadays :-)

File: src/fastfilter.h (hunk 4 above)
Comment (awemany) [reply src/fastfilter.h:143]: I am worried about interaction of `roll()` and an adversary.

File: src/test/fastfilter_tests.cpp (hunk 5 above)
Comment (awemany) [reply src/test/fastfilter_tests.cpp:40]: Ok, that's good enough!

File: src/fastfilter.h (hunk 8 above)
Comment (awemany) [reply src/fastfilter.h:100]: But the key here is `all` the bits. Do you think this will be faster (by a worthwhile amount) compared to just proper slicing?

**Comment by awemany** [comment]
@gandrewstone: Awesome wrt. the bench test.  I still need to take a closer look, though.

File: src/fastfilter.h (hunk 2 above)
Comment (gandrewstone) [reply src/fastfilter.h:105]: They aren't.  I could probably get another 2 to 10x improvement if I used assembly.  This is why libsecp is asm.  Note also that inlining is optional in the compiler.

File: src/fastfilter.h (hunk 4 above)
Comment (gandrewstone) [reply src/fastfilter.h:143]: That adversary would only penalize his own tx.  Additionally in looking at the current bloom implementation, I've realized that it's hash is not cryptograpically strong.  So you'd have to prove it's properties before making these differences  claims.

File: src/fastfilter.h (hunk 8 above)
Comment (gandrewstone) [reply src/fastfilter.h:100]: All the bits are used if you choose all the hashes.  Which we do.  If you choose fewer hashes, you are making a tradeoff of performance verses deliberate collision attacks IF and only if this structure is used with data specified by an external source.  Even in that case, what are the repercussions of collisions?  Could be minimal.  I feel like you are rejecting for example the radix tree data structure because it's possible to insert data that turns that tree into a list.  But there are many applications of radix trees where we have strong knowledge that the data won't do that.

Comment (sickpig) [reply src/fastfilter.h:100]: unrelated point: we sure that `__builtin_bswap32` compiler functions  is always available?

File: src/bench/bloom.cpp [hunk 9]
```diff
@@ -0,0 +1,192 @@
+// Copyright (c) 2016 The Bitcoin Core developers
```
Comment (sickpig) [inline src/bench/bloom.cpp]: Both date and Copyright owner seem not correct.

File: src/bloom.cpp [hunk 10]
```diff
@@ -44,7 +44,6 @@ void CBloomFilter::setup(unsigned int nElements,
     isEmpty = true;
 
     nHashFuncs = (unsigned int)(vData.size() * 8 / nElements * LN2);
-
```
Comment (sickpig) [inline src/bloom.cpp]: nit: unrelated new line

**Comment by sickpig** [comment]
@gandrewstone need to be rebased

**Comment by ptschip** [comment]
needs a rebase, conflicts with bloom.cpp

File: src/fastfilter.h (hunk 8 above)
Comment (gandrewstone) [reply src/fastfilter.h:100]: It's a gcc extension, so its not always available.  However, many compilers support gcc extensions, and its compiling on all of our architectures.  If we port to an unsupported compiler we will need to add a compatibility macro at that time.

File: src/bench/bloom.cpp (hunk 9 above)
Comment (gandrewstone) [reply src/bench/bloom.cpp]: thanks, fixed

**Comment by gandrewstone** [comment]
rebased, fixed copyright

**Comment by awemany** [comment]
@gandrewstone:

Wrt. this comment part:
```
// Since the false positive rate is the chance that a random value insertion hits one already there, it
// is the same as the fill ratio. 

```

Do I misunderstand it or is it incorrect? The false positive rate is that there is a hit at all 16 locations, which is different from how you define the 'fill ratio' - or am I missing something?

**Comment by awemany** [comment]
Another remark / change request: You bitwise-and with `FILTER_SIZE - 1` to stay within the filter's array bounds . As far as I can see, there's no code that checks (like a `static_assert`) whether this is actually a power of two. This might lead to a potential out of-bounds access problem elsewhere in the code!

EDIT: Remark #2: The `REL_PRIME` value does not seem to be used anywhere.

**Comment by gandrewstone** [comment]
@awemany done.  I think also that you will appreciate the additional details I put into the description.

File: src/fastfilter.h [hunk 11]
```diff
@@ -23,17 +23,25 @@ class uint256;
...
+ * execution will be faster).  Therefore, if this structure is used in an application that accepts externally created
+ * uint256 numbers that are sensitive to deliberately constructed collisions, be sure to keep NUM_HASH_FNS high enough
+ * that the creation of collisions in the used bits is not feasible.
+ *
+ * This class is thread-safe in the sense that simultaneous calls to member functions will not crash,
  * but "inserts" may be lost.  However, if you are using this class as an in-ram filter before doing a more expensive
  * operation, a lost insert may be acceptable.
+ *
+ * FILTER_SIZE must be a power of 2, and NUM_HASH_FNs may range from 2 to 16 inclusive.  Since hashes are calculated
+ * in pairs of 2, odd values of NUM_HASH_FNs are rounded down.
```
Comment (awemany) [inline src/fastfilter.h]: Nit: NUM_HASH_FN**S**

File: src/fastfilter.h (hunk 4 above)
Comment (awemany) [reply src/fastfilter.h:143]: > That adversary would only penalize his own tx.

Yes, if the fast filter is only used to filter out transactions from a peer. My worry is that its usage grows and implicit assumptions about its hardness are made and at some point we're open to an attack.

> Additionally in looking at the current bloom implementation, I've realized that it's hash is not cryptograpically strong.

As far as I can see, it is not the cryptographically strong that is important here, rather that the hash function is unpredictable from the attacker's POV (e.g. how siphash is described here: https://en.wikipedia.org/wiki/SipHash) 

I don't know whether murmurhash in `CBloomFilter` has that property.

I guess this could all be subsumed under 'be careful when using this' and corresponding warnings in the code.

**Review by awemany — CHANGES_REQUESTED** [review]
Another minor nit.

**Comment by gandrewstone** [comment]
unpredictability of the output given an input is one of the required properties for a hash function to be considered cryptographically strong so my statement is not incorrect.

Optimization often requires that special-purpose data structures be created that may not be applicable in other contexts.  If we do not allow data structures because they may be misused in other contexts we won't get very far.  I've added words describing the issues in the documentation of fastfilter.h.

I'm hoping to get an approval from you soon so this can be merged.  I don't want to delay this for another day over minor nits in comments.

File: src/fastfilter.h [hunk 12]
```diff
@@ -23,17 +23,25 @@ class uint256;
  *
  * This class can be used anywhere a Bloom filter is used so long as the input data is random.
  *
- * This class is thread-safe in the sense that simultaneous calls to insert and contains will not crash,
+ * If NUM_HASH_FNS is 16 and FILTER_SIZE is >= 64k all bits in the uint256 input data will be used to set bits in
+ * the filter.  If these fields are set to lower numbers, fewer bits may be used (although in the NUM_HASH_FNS case
+ * execution will be faster).  Therefore, if this structure is used in an application that accepts externally created
```
Comment (awemany) [inline src/fastfilter.h:28]: I think this should be further clarified a bit. It is not only that 'all bits are used' it is that they are **re**used in a straight-forward redundant way without further mixing when the filter size is >64k, which will very likely drive up false positive rates above what would be expected by a proper bloom approach. The user should know this.

**Review by awemany — COMMENTED** [review]
@gandrewstone: Fair enough. I don't want to needlessly stand in the way here though I would really prefer a simpler to analyze straight-forward hashing (though maybe slightly less performant) approach.  This feels like premature optimization that will bite us. However, there's another comment I just added to the docs section so that any user at least knows what he's dealing with in a bit more detail.

EDIT: Maybe a further comment here on the soundness in terms of security of the fast filter. I think the expectation that:

- with less than 65536 entries and
-  n=16 hash functions
- a check of the assumption that the input data is a cryptographic hash

 the fast filter behaves like a bloom filter under attack that has the given number of entries but only `n=16 - <number-of-bits-controlled-by-attacker / 16>` effective hash functions seems to be sound. I am fully o.k. with that use case.

For all other values, the analysis of the fast filter is lacking and I feel wary about its use.

On this:

> unpredictability of the output given an input is one of the required properties for a hash function to be considered cryptographically strong so my statement is not incorrect.

Your statement is not incorrect but it would miss the point - as we're talking about its negation. The question I raised is whether `FastFilter` would have a larger attack surface than `CBloomFilter` which depends upon the noted property or lack thereof in murmurhash. It might well be the latter, I didn't check.

File: src/fastfilter.h (hunk 12 above)
Comment (gandrewstone) [reply src/fastfilter.h:28]: done

File: src/fastfilter.h [hunk 13]
```diff
@@ -29,12 +29,18 @@ constexpr bool isPow2(unsigned int num) { return num && !(num & (num - 1)); }
  * uint256 numbers that are sensitive to deliberately constructed collisions, be sure to keep NUM_HASH_FNS high enough
  * that the creation of collisions in the used bits is not feasible.
  *
+ * Note also that the input bits are used without obfuscation or mixing so if an attacker can control some input bits
+ * the attacker can cause collisions in some of the fast filter entries for his inputs.  This will cause higher false
+ * positive rates for these transactions.  For example, if the attacker can control 32 bits of the input, he can
+ * effectively reduce the number of hash functions in the fast filter by 2 because he has engineered a guaranteed
+ * collision for the two functions that use those bits.
+ *
```
Comment (awemany) [inline src/fastfilter.h:37]: Looks good!

I'd still add a comment along the lines of 'for >64k entries and << 16 hash functions, there might be dragons' so that everyone knows that this would be outside the 'safe operating area' of the fast filter.

Comment (gandrewstone) [reply src/fastfilter.h:37]: I've got those numbers in the paragraph above.  To expedite this, can you write whatever sentence works for you in a comment and I'll put it in if I agree with it?

Comment (gandrewstone) [reply src/fastfilter.h:37]: We need to turn around comment->change faster than once per day, especially when we are just dealing with comments -- and potentially short lived comment at that,  since they may be gone as soon as we find the time to invent or discover an efficient algorithm that obfuscates input bits.  Delaying this has already wasted @ptschip time because some collision reduction code in another PR was expecting the collision reduction in this one.  

Alternately,  lets get this approved and merged and we can open a new PR reworking the comments and any other small things.

**Review by ptschip — APPROVED** [review]
After reviewing and doing testing with this I think this looks very good.  The issue of the false positives I was reporting has nothing to do with the fast filter but has only to do with the bytes chosen before getting xor'd and *before* insertion into the filter.

File: src/fastfilter.h (hunk 13 above)
Comment (awemany) [reply src/fastfilter.h:37]: EDIT: I opened PR #1373 as a continuation of this review comment thread.
